### PR TITLE
Make runtime benchmark harness more flexible to use

### PR DESCRIPTION
**Pull request by Kobzol**
Some runtime benchmarks need only read-only access to some input data on which they operate on. Currently, it wasn't easily possible to express this in the runtime benchmark harness, and even though a benchmark only needed to read from some data, the data had to be re-created for each iteration of the benchmark (twice actually, since we run each benchmark two times, to gather instructions and time).

This meant that the data for each benchmark had to be re-created many times, which could slow down the execution of the whole benchmark group. And also the data was always created anew, possibly on a new memory address, which could introduce more noise into the benchmark, because each iteration of the benchmark was operating potentially on different regions of memory.

After this change, benchmarks that only need read-only access to some input data should only create that data once and then read it repeatedly. This can in theory mean that the data will be more available in L1/L2/L3 caches, but that's an orthogonal problem. It is even a question whether we want to avoid this, we could instead warm up the cache and perform measurements on a "prepared cache state", which could further reduce measurement variation. In any case, I have some code prepared to flush the caches if this would be considered a problem.

### CODE REVIEW

**Comment by Kobzol** [comment]
One slightly annoying thing is that the data will be prepared even if the runtime binary is just executed with the `list` parameter, which returns a list of benchmarks contained in the binary. I'm not sure what to do about that though. We only use the `list` command for nicer CLI output and duplicate benchmark detection, we could gather the benchmark names only when we actually run the runtime benchmark binary instead..

File: collector/benchlib/src/benchmark.rs [hunk 1]
```diff
@@ -7,30 +7,26 @@ use std::collections::HashMap;
 
 /// Create and run a new benchmark group. Use the closure argument to register
 /// the individual benchmarks.
-pub fn run_benchmark_group<F>(register: F)
+pub fn run_benchmark_group<'scope, F>(register: F)
```
Comment (nnethercote) [inline collector/benchlib/src/benchmark.rs:10]: Rename `'scope` as `'a`? It's much shorter, and `'scope` isn't informative.

**Review by nnethercote — APPROVED** [review]
> One slightly annoying thing is that the data will be prepared even if the runtime binary is just executed with the list parameter, which returns a list of benchmarks contained in the binary.

Should this be mentioned in a comment somewhere?

**Comment by Kobzol** [comment]
Thanks, I have expanded the `README` of the runtime benchmarks directory.

**Pull request merged**